simulation/chart.py

```diff
@@ -1,232 +1,427 @@
-    legend = chart.addLegend(offset=None)
-    vb: pg.ViewBox = layout.addViewBox()
-    if legend_width is not None:
-        vb.setFixedWidth(legend_width)
-    legend.setParentItem(vb)
-    legend.anchor((0, 0), (0, 0))
-    chart.setClipToView(True)
 
-    if x_label is not None:
-        chart.setLabel("bottom", x_label)
-    if x_log is not None:
-        chart.setLogMode(x=x_log)
-    if x_range is not None:
-        chart.setXRange(*x_range)
+    index: int
+    hues: int
 
-    if y_label is not None:
-        chart.setLabel("left", y_label)
-    if y_log is not None:
-        chart.setLogMode(y=y_log)
-    if y_range is not None:
-        chart.setYRange(*y_range)
 
-    layout.nextRow()
+Color = Union[RGBAColor, NamedColor, IndexedColor, int, float, str]
+"""Color based on model, predefined values, single greyscale value (0.0 - 1.0) , color index or string representing color (“#RGB”,“#RGBA”,“#RRGGBB”,“#RRGGBBAA”)."""
 
-    return chart
 
+@singledispatch
+def _make_color(color: RGBAColor):  # pyright:ignore [reportGeneralTypeIssues]
+    """Convert RGBA color value to QColor"""
+    if color.A is None:
+        return pg.mkColor((color.R, color.G, color.B))
+    else:
+        return pg.mkColor((color.R, color.G, color.B, color.A))
 
-def add_line_plot(
-    chart: pg.PlotItem, name: str, color=None, width: Optional[float] = None, style=None
-) -> pg.PlotDataItem:
-    """Adds a line plot to the chart.
 
-    Args:
-        name (str): Name (legend title) of the plot
-        color (Any, optional): Line color of the plot. Defaults to None.
-        width (float, optional): Width of the plot line. Defaults to None.
-        style (Any, optional): Style of the plot line. Defaults to None.
-    """
-    config = {}
-    if color is not None:
-        config["color"] = color
+@singledispatch
+def _make_color(color: IndexedColor):  # pyright:ignore [reportGeneralTypeIssues]
+    """Convert Indexed color value to QColor"""
+    return pg.mkColor((color.index, color.hues))
+
+
+@singledispatch
+def _make_color(color: Union[int, float, str, None]):
+    if color is None:
+        return None
+    return pg.mkColor(color)
+
+
+class PenStyle(Enum):
+    """Pen style for drawing lines"""
+
+    # NoPen                    = QtCore.Qt.PenStyle.NoPen
+    SolidLine = QtCore.Qt.PenStyle.SolidLine  # pyright:ignore [reportGeneralTypeIssues]
+    DashLine = QtCore.Qt.PenStyle.DashLine  # pyright:ignore [reportGeneralTypeIssues]
+    DotLine = QtCore.Qt.PenStyle.DotLine  # pyright:ignore [reportGeneralTypeIssues]
+    DashDotLine = (
+        QtCore.Qt.PenStyle.DashDotLine  # pyright:ignore [reportGeneralTypeIssues]
+    )
+    DashDotDotLine = (
+        QtCore.Qt.PenStyle.DashDotDotLine  # pyright:ignore [reportGeneralTypeIssues]
+    )
+    # CustomDashLine           = QtCore.Qt.PenStyle.CustomDashLine
+    # MPenStyle                = QtCore.Qt.PenStyle.MPenStyle
+
+
+def _make_pen(
+    color: Optional[Color] = None,
```

Comment:
Just look for all cases in this file where you do some sort of check for `if variable is (not) None`. See if you can replace these using sensible defaults.

Reply:
https://github.com/semiotic-ai/autoagora-agents/pull/21/commits/0d488125217f7472c649d43cfd26daae40c6cbec

Reply:
Still lots of `None` checking.

Reply:
The remaining `None` checks are necessary to take action or not. for example if output_file is defined this means we expect to record a video otherwise just show the GUI. Or `y_range` which defines range otherwise automatic.

Reply:
Gotcha. In this case, in my opinion, a better pattern would be to reflect the style of the library we're wrapping and create a separate `set_yrange` function that users could call. In the interest of time, I'm happy to instead say that we just clean up this `None` RTTI a bit by compressing to one line. You can run the following as an example.

```python
a = "I am set!"

def foo():
    print("I was run!")

a and foo()  # Runs foo

a = None

a and foo()  # Won't run foo
```

Reply:
removed all ` if value is not None:`
https://github.com/semiotic-ai/autoagora-agents/pull/21/commits/09de1ed65bd4153fbcdd7759e11d3dbb3e4a0155

Reply:
Thanks!